README.md

```diff
@@ -109,6 +103,124 @@ from `run_benchmark()` (both default `FALSE`):
+```
+new_csv_benchmark <- Benchmark(
+  "new_csv_benchmark",
+  # This setup block will be run before the benchmark is started. This is run
+  # before each case / single item in the benchmark matrix, so is a good time
+  # to setup case- or source-specific properties (see `result_dim` below).
+  setup = function(source = names(known_sources),
+                   as_data_frame = c(TRUE, FALSE),
+                   skip_empty_rows = TRUE)) {
+    # Validate the parameters
+    # For our benchmark: as_data_frame defaults to TRUE and FALSE (so if it is 
+    # unspecified you will get both TRUE and FALSE in the benchmark matrix)
+    as_data_frame <- match.arg(as_data_frame)
+
+    # For our benchmark: skip_empty_rows defaults to TRUE (so if it is 
+    # unspecified you will only get TRUE in the benchmark matrix), but it can 
+    # accept TRUE or FALSE, so we validate that it is one of those.
+    skip_empty_rows <- match.arg(skip_empty_rows, c(TRUE, FALSE))
+
+    # Ensure the file exists as an uncompressed csv
+    input_file <- ensure_format(source, "csv", "uncompressed")
+
+    # Extract the dim attribute from a data source for validation later
+    result_dim <- get_source_attr(source, "dim")
+
+    # Finally we return a `BenchEnvironment` with the parameters we defined 
+    # above that are needed
+    BenchEnvironment(
+      input_file = input_file,
+      result_dim = result_dim,
+      as_data_frame = as_data_frame,
+      skip_empty_rows = skip_empty_rows
+    )
+  },
+  # This is run before each iteration. 
+  before_each = {
+    # Make sure the result is cleared
+    result <- NULL
+  },
+  # This is the only part of the code that is actually measured when the benchmark 
+  # is run. It should include all and only the code you are interested in benchmarking.
+  run = {
+    result <- read_csv_arrow(
+      input_file, 
+      as_data_frame = as_data_frame, 
+      skip_empty_rows = skip_empty_rows
+    )
+  },
+  # This is run after each iteration. This is a good time to validate that the
+  # benchmark ran correctly.
+  after_each = {
+    stopifnot(
+      "The dimensions do not match" = all.equal(dim(result), result_dim)
+    )
+    result <- NULL
+  },
+  # This defines if the parameters are valid. If there are certain combinations
+  # that are not valid, add them to drop and they will be excluded from the 
+  # benchmark matrix
+  valid_params = function(params) {
+  
+    # Do not allow both skip_empty_rows == FALSE and as_data_frame == TRUE at 
+    # the same time
+    drop <- ( params$skip_empty_rows == FALSE & params$as_data_frame == TRUE ) 
+    
+    params[!drop,]
+  },
+  # This lists any packages that are used by this benchmark so that they can 
+  # be installed prior to starting the run. Typically this will be simply "arrow"
+  packages_used = function(params) {
+    "arrow"
+  }
+)
+```
+
+And now we could run our benchmark with the following for the default matrix, 
+using 4 cpu cores and 5 iterations per case.
```

Comment:
Should explain "iterations" somewhere, probably above under "Running benchmarks"